src/librustc/ty/structural_match.rs

```diff
@@ -207,3 +301,186 @@ impl<'a, 'tcx> TypeVisitor<'tcx> for Search<'a, 'tcx> {
+                                impose_partialeq_bound = false;
+                            }
+                        } else {
+                            self.infcx.tcx.sess.delay_span_bug(self.span, &format!(
+                                "SearchHirExpr didn't resolve def_id: {:?}", def_id));
+                        }
+                    }
+                    Res::Def(DefKind::Ctor(..), _def_id) => {
+                        self.register_structural_bound(ty);
+                    }
+                    _ => {
+                        debug!("SearchHirExpr ExprKind::Path res: {:?} \
+                                traverse type instead", res);
+                    }
+                }
+            }
+
+            hir::ExprKind::Box(..) |
+            hir::ExprKind::Array(..) |
+            hir::ExprKind::Repeat(..) |
+            hir::ExprKind::Tup(..) |
+            hir::ExprKind::Type(..) |
+            hir::ExprKind::DropTemps(..) |
+            hir::ExprKind::AddrOf(..) => {
+                // continue expression-based traversal
+                intravisit::walk_expr(self, ex)
+            }
+
+            hir::ExprKind::Block(block, _opt_label) => {
+                // skip the statements, focus solely on the return expression
+                if let Some(ex) = &block.expr {
+                    intravisit::walk_expr(self, ex)
+                }
+            }
+            hir::ExprKind::Match(_input, arms, _match_source) => {
+                // skip the input, focus solely on the arm bodies
+                for a in arms.iter() {
+                    intravisit::walk_expr(self, &a.body)
+                }
+            }
+
+            hir::ExprKind::Index(base, _index) => {
+                // skip the index, focus solely on the base content.
+                // (alternative would be to do type-based analysis, which would
+                // be even more conservative).
+                intravisit::walk_expr(self, base);
+            }
+
+            hir::ExprKind::Loop(..) |
+            hir::ExprKind::Call(..) |
+            hir::ExprKind::MethodCall(..) |
+            hir::ExprKind::Binary(..) |
+            hir::ExprKind::Unary(..) |
+            hir::ExprKind::Cast(..) |
+            hir::ExprKind::Closure(..) |
+            hir::ExprKind::Assign(..) |
+            hir::ExprKind::AssignOp(..) |
+            hir::ExprKind::Field(..) |
+            hir::ExprKind::Break(..) |
+            hir::ExprKind::Continue(..) |
+            hir::ExprKind::Ret(..) |
+            hir::ExprKind::Yield(..) |
+            hir::ExprKind::InlineAsm(..) |
+            hir::ExprKind::Lit(..) |
+            hir::ExprKind::Err => {
+                // abstraction barrier for non-trivial expression; traverse
+                // `typeof(expr)` instead of expression itself.
+                debug!("SearchHirExpr switch to type analysis for expr: {:?} ty: {:?}", ex, ty);
+                self.search_for_structural_match_violation(ty);
```

Comment:
Heh, this is a lot like const-qualification (cc @oli-obk @RalfJung @ecstatic-morse).

Maybe one day we'll have an unified system for refining a trait bound based on values.